src/CodeGen_LLVM.cpp

```diff
@@ -426,63 +520,72 @@ void CodeGen_LLVM::compile(Stmt stmt, string name,
     // Now verify the function is ok
     internal_assert(verifyFunction(*function) == false);
 
-    // Now we need to make the wrapper function (useful for calling from jit)
-    string wrapper_name = name + "_argv";
-    func_t = FunctionType::get(i32, vec<llvm::Type *>(i8->getPointerTo()->getPointerTo()), false);
-    llvm::Function *wrapper = llvm::Function::Create(func_t, llvm::Function::ExternalLinkage, wrapper_name, module);
-    block = BasicBlock::Create(*context, "entry", wrapper);
-    builder->SetInsertPoint(block);
-
-    Value *arg_array = wrapper->arg_begin();
+    // If the Func is externally visible, also create the argv wrapper
+    // (useful for calling from JIT and other machine interfaces).
+    if (op->linkage == LoweredFunc::External) {
+        add_argv_wrapper(module, function, name + "_argv");
+    }
+}
 
-    vector<Value *> wrapper_args(args.size());
-    for (size_t i = 0; i < args.size(); i++) {
-        // Get the address of the nth argument
-        Value *ptr = builder->CreateConstGEP1_32(arg_array, (int)i);
-        ptr = builder->CreateLoad(ptr);
-        if (args[i].is_buffer()) {
-            // Cast the argument to a buffer_t *
-            wrapper_args[i] = builder->CreatePointerCast(ptr, buffer_t_type->getPointerTo());
-        } else {
-            // Cast to the appropriate type and load
-            ptr = builder->CreatePointerCast(ptr, arg_types[i]->getPointerTo());
-            wrapper_args[i] = builder->CreateLoad(ptr);
-        }
+// Given a range of iterators of constant ints, get a corresponding vector of llvm::Constant.
+template<typename It>
+std::vector<llvm::Constant*> get_constants(llvm::Type *t, It begin, It end) {
+    std::vector<llvm::Constant*> ret;
+    for (It i = begin; i != end; i++) {
+        ret.push_back(ConstantInt::get(t, *i));
     }
-    debug(4) << "Creating call from wrapper to actual function\n";
-    Value *result = builder->CreateCall(function, wrapper_args);
-    builder->CreateRet(result);
-    internal_assert(verifyFunction(*wrapper) == false);
+    return ret;
+}
 
-    // Finally, verify the module is ok
-    internal_assert(verifyModule(*module) == false);
-    debug(2) << "Done generating llvm bitcode\n";
+void CodeGen_LLVM::visit(const Buffer *op) {
```

Comment:
Yeah, this could change. At first, I was trying to make it so Buffer (and even LoweredFunc) could exist as non-root IR nodes (e.g. declaring a Buffer as a local function object), and they were IR nodes when I started working on this. But, obviously that isn't something that works now. 

I will change these.